browser/tor/tor_control.cc

```diff
@@ -0,0 +1,1512 @@
+  }
+
+  // Get the file's info, including modification time.
+  base::File::Info info;
+  if (!portfile.GetInfo(&info)) {
+    VLOG(0) << "tor: failed to stat control port";
+    return false;
+  }
+
+  // Read up to 27/28 octets, the maximum we will ever need.
+#if defined(OS_WIN)
+  const char mintmpl[] = "PORT=1.1.1.1:1\r\n";
+  const char maxtmpl[] = "PORT=255.255.255.255:65535\r\n";
+#else
+  const char mintmpl[] = "PORT=1.1.1.1:1\n";
+  const char maxtmpl[] = "PORT=255.255.255.255:65535\n";
+#endif
+  const size_t kBufSiz = strlen(maxtmpl);
+  char buf[kBufSiz];
+  int nread = portfile.ReadAtCurrentPos(buf, sizeof buf);
+  if (nread < 0) {
+    VLOG(0) << "tor: failed to read control port";
+    return false;
+  }
+  if (static_cast<size_t>(nread) < strlen(mintmpl)) {
+    VLOG(0) << "tor: control port truncated";
+    return false;
+  }
+  DCHECK(static_cast<size_t>(nread) <= sizeof buf);
+
+  std::string text(buf, 0, nread);
+
+  // Sanity-check the content.
+  if (!base::StartsWith(text, "PORT=", base::CompareCase::SENSITIVE) ||
+#if defined(OS_WIN)
+      !base::EndsWith(text, "\r\n", base::CompareCase::SENSITIVE)) {
+#else
+      !base::EndsWith(text, "\n", base::CompareCase::SENSITIVE)) {
+#endif
+    VLOG(0) << "tor: invalid control port: "
+               << "`" << text << ";";  // XXX escape
+    return false;
+  }
+
+  // Verify that it's localhost.
+  const char expected[] = "PORT=127.0.0.1:";
+  if (!base::StartsWith(text, expected, base::CompareCase::SENSITIVE)) {
+    VLOG(0) << "tor: control port has non-local control address";
+    return false;
+  }
+
+  // Parse it!
+#if defined(OS_WIN)
+  std::string portstr(text, strlen(expected), nread - 2 - strlen(expected));
+#else
+  std::string portstr(text, strlen(expected), nread - 1 - strlen(expected));
+#endif
+  if (!base::StringToInt(portstr, &port)) {
+    VLOG(0) << "tor: failed to parse control port: "
+               << "`" << portstr << "'";  // XXX escape
+    return false;
+  }
+  mtime = info.last_modified;
+  VLOG(3) << "Control port " << port << ", mtime " << mtime;
+  return true;
+}
+
+bool TorControl::EatOldPid(base::ProcessId* id) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(watch_sequence_checker_);
+  DCHECK(id);
+
+  // Open the tor pid file.
+  base::FilePath pidpath = watch_dir_path_.AppendASCII("tor.pid");
+  base::File pidfile(pidpath,
+                        base::File::FLAG_OPEN|base::File::FLAG_READ);
+  if (!pidfile.IsValid()) {
+    VLOG(0) << "tor: failed to open tor.pid";
+    return false;
+  }
+
+#if defined(OS_WIN)
+  const char maxpid[] = "4294967292";  // 0xFFFFFFFC
+#else
+  const char maxpid[] = "4194304";  // 0x400000
+#endif
+  const size_t kBufSiz = strlen(maxpid);
+  char buf[kBufSiz];
+  int nread = pidfile.ReadAtCurrentPos(buf, sizeof buf);
+  if (nread < 0) {
+    VLOG(0) << "tor: failed to read tor pid file";
+    return false;
+  }
+
+  std::string pid(buf, 0, nread - 1);
+#if defined(OS_WIN)
+  *id = stoul(pid, nullptr);
```

Comment:
because of the definition of `base::ProcessId`
```
#if defined(OS_WIN)
typedef HANDLE ProcessHandle;
typedef DWORD ProcessId;
typedef HANDLE UserTokenHandle;
const ProcessHandle kNullProcessHandle = NULL;
const ProcessId kNullProcessId = 0;
#elif defined(OS_FUCHSIA)
typedef zx_handle_t ProcessHandle;
typedef zx_koid_t ProcessId;
const ProcessHandle kNullProcessHandle = ZX_HANDLE_INVALID;
const ProcessId kNullProcessId = ZX_KOID_INVALID;
#elif defined(OS_POSIX)
// On POSIX, our ProcessHandle will just be the PID.
typedef pid_t ProcessHandle;
typedef pid_t ProcessId;
const ProcessHandle kNullProcessHandle = 0;
const ProcessId kNullProcessId = 0;
#endif  // defined(OS_WIN)
```